### PR TITLE
port to github/pierrec/lz4/v4

### LINKED ISSUE #248: RFE: port to github.com/pierrec/lz4/v4
## What would you like to have changed?
Make it possible to build with `github.com/pierrec/lz4/v4`.

## Why is this feature a useful, necessary, and/or important addition to this project?
`lz4` was updated to 4.0.2 in Fedora rawhide and `archiver` no longer builds using distribution-provided Go packages as a result.

## What alternatives are there, or what are you doing in the meantime to work around the lack of this feature?
I'm forced to stop updating `archiver` until it's ported to `lz4` v4 API or someone creates a compatibility package with v3 API.

## Please link to any relevant issues, pull requests, or other discussions.

To see the issue, patch the source with the following patch:
``` diff
diff -up archiver-3.3.2/lz4.go.lz4 archiver-3.3.2/lz4.go
--- archiver-3.3.2/lz4.go.lz4	2020-09-28 10:43:21.000000000 +0200
+++ archiver-3.3.2/lz4.go	2020-10-05 13:06:07.879465436 +0200
@@ -5,7 +5,7 @@ import (
 	"io"
 	"path/filepath"
 
-	"github.com/pierrec/lz4/v3"
+	"github.com/pierrec/lz4"
 )
 
 // Lz4 facilitates LZ4 compression.
diff -up archiver-3.3.2/tarlz4.go.lz4 archiver-3.3.2/tarlz4.go
--- archiver-3.3.2/tarlz4.go.lz4	2020-09-28 10:43:21.000000000 +0200
+++ archiver-3.3.2/tarlz4.go	2020-10-05 13:06:17.578418304 +0200
@@ -5,7 +5,7 @@ import (
 	"io"
 	"strings"
 
-	"github.com/pierrec/lz4/v3"
+	"github.com/pierrec/lz4"
 )
 
 // TarLz4 facilitates lz4 compression
```
and build. Fedora package build process fails with the following errors:
```
_build/src/github.com/mholt/archiver/lz4.go:19:3: w.Header undefined (type *lz4.Writer has no field or method Header)
_build/src/github.com/mholt/archiver/tarlz4.go:87:7: lz4w.Header undefined (type *lz4.Writer has no field or method Header)
```

It looks like option handling and compression level setting got changed:
https://github.com/pierrec/lz4/compare/v3.3.2..v4.0.2

I have nearly zero Golang knowledge, but this crude patch makes it compile and `go test` passes:
``` diff
diff -up archiver-3.3.2/lz4.go.lz4 archiver-3.3.2/lz4.go
--- archiver-3.3.2/lz4.go.lz4	2020-09-28 10:43:21.000000000 +0200
+++ archiver-3.3.2/lz4.go	2020-10-05 13:28:21.581995885 +0200
@@ -5,7 +5,7 @@ import (
 	"io"
 	"path/filepath"
 
-	"github.com/pierrec/lz4/v3"
+	"github.com/pierrec/lz4"
 )
 
 // Lz4 facilitates LZ4 compression.
@@ -16,7 +16,12 @@ type Lz4 struct {
 // Compress reads in, compresses it, and writes it to out.
 func (lz *Lz4) Compress(in io.Reader, out io.Writer) error {
 	w := lz4.NewWriter(out)
-	w.Header.CompressionLevel = lz.CompressionLevel
+	options := []lz4.Option{
+		lz4.CompressionLevelOption(lz4.CompressionLevel(1 << (8 + lz.CompressionLevel))),
+	}
+	if err := w.Apply(options...); err != nil {
+		return err
+	}
 	defer w.Close()
 	_, err := io.Copy(w, in)
 	return err
diff -up archiver-3.3.2/tarlz4.go.lz4 archiver-3.3.2/tarlz4.go
--- archiver-3.3.2/tarlz4.go.lz4	2020-09-28 10:43:21.000000000 +0200
+++ archiver-3.3.2/tarlz4.go	2020-10-05 13:28:21.581995885 +0200
@@ -5,7 +5,7 @@ import (
 	"io"
 	"strings"
 
-	"github.com/pierrec/lz4/v3"
+	"github.com/pierrec/lz4"
 )
 
 // TarLz4 facilitates lz4 compression
@@ -84,7 +84,12 @@ func (tlz4 *TarLz4) wrapWriter() {
 	var lz4w *lz4.Writer
 	tlz4.Tar.writerWrapFn = func(w io.Writer) (io.Writer, error) {
 		lz4w = lz4.NewWriter(w)
-		lz4w.Header.CompressionLevel = tlz4.CompressionLevel
+		options := []lz4.Option{
+			lz4.CompressionLevelOption(lz4.CompressionLevel(1 << (8 + tlz4.CompressionLevel))),
+		}
+		if err := lz4w.Apply(options...); err != nil {
+			return lz4w, err
+		}
 		return lz4w, nil
 	}
 	tlz4.Tar.cleanupWrapFn = func() {
```

### PR DESCRIPTION
**Pull request by rathann**
Resolves #248 .

Minimum lang version needs to be bumped to 1.13, otherwise we get this error:
```
./lz4.go:20:53: invalid operation: 1 << (8 + lz.CompressionLevel) (signed shift count type int) requires go1.13 or later (-lang was set to go1.12; check go.mod)
./tarlz4.go:88:54: invalid operation: 1 << (8 + tlz4.CompressionLevel) (signed shift count type int) requires go1.13 or later (-lang was set to go1.12; check go.mod)
```

### CODE REVIEW

File: tarlz4.go [hunk 1]
```diff
@@ -84,7 +84,12 @@ func (tlz4 *TarLz4) wrapWriter() {
 	var lz4w *lz4.Writer
 	tlz4.Tar.writerWrapFn = func(w io.Writer) (io.Writer, error) {
 		lz4w = lz4.NewWriter(w)
-		lz4w.Header.CompressionLevel = tlz4.CompressionLevel
+		options := []lz4.Option{
+			lz4.CompressionLevelOption(lz4.CompressionLevel(1 << (8 + tlz4.CompressionLevel))),
```
Comment (coolaj86) [inline tarlz4.go:88]: This should be `lz4.CompressionLevelOption(lz4.Level1)`

File: lz4.go [hunk 2]
```diff
@@ -16,7 +16,12 @@ type Lz4 struct {
 // Compress reads in, compresses it, and writes it to out.
 func (lz *Lz4) Compress(in io.Reader, out io.Writer) error {
 	w := lz4.NewWriter(out)
-	w.Header.CompressionLevel = lz.CompressionLevel
+	options := []lz4.Option{
+		lz4.CompressionLevelOption(lz4.CompressionLevel(1 << (8 + lz.CompressionLevel))),
```
Comment (coolaj86) [inline lz4.go:20]: Same comment here as below. Default compression should not need to be set, but if we want to set it to be explicit (which I think we do), it should be `lz4.Fast`.

Comment (coolaj86) [reply lz4.go:20]: Ah, I see why this was done. This way does preserve backwards compatibility, but I think it deserves a comment, something like:

```
// TODO archiver v4: use proper lz4.Fast
// bitshifting for backwards compatibility with lz4/v3
```

**Comment by coolaj86** [comment]
From what I can tell it looks like this is safe to merge without breaking backwards compatibility.

File: lz4.go (hunk 2 above)
Comment (rathann) [reply lz4.go:22]: Done,

**Comment by rathann** [comment]
I'm fine with releasing this only as part of v4. It's just that I won't be able to rebuild `archiver` in Fedora development branch (rawhide) until either this is merged or someone packages `lz4/v3` compatibility package in Fedora.

**Review by coolaj86 — APPROVED** [review]
I don't see any breaking changes with the internal implementation details.

I think updating to go1.13 is reasonable given that we're at go1.15 now and, as always, there are no forwards-incompatible changes.

Will test locally and merge.

**Pull request merged**